### PR TITLE
Restrict repository indexing by glob match

### PR DESCRIPTION
**Pull request by guillep2k**
### EDIT: This PR mutated along the way; to get the current state of this PR, please jump to [this comment](https://github.com/go-gitea/gitea/pull/7767#issuecomment-528701152).

Add an option in `app.ini` to restrict repository indexing by file extension:
```
[Indexer]
...
REPO_INDEXER_EXTENSIONS = go, c, cpp, h, idl, php, .
REPO_EXTENSIONS_LIST_EXCLUDE = false
```
It's not very powerful, but can help when indexes become polluted with derived files (e.g. `lss`, `sym`, `map`, etc.) so the search results are more relevant.

**REPO_INDEXER_EXTENSIONS** (default: empty) is a comma separated list of file extensions to include/exclude from the index. Only simple strings are supported (e.g. can't exclude `,v` from CVS). A dot (`.`) matches files with no extension. An empty list means "include all files", disregarding the value of **REPO_EXTENSIONS_LIST_EXCLUDE** (i.e. the current behaviour).
**REPO_EXTENSIONS_LIST_EXCLUDE** (default: `false`) defines whether **REPO_INDEXER_EXTENSIONS** is a list of extensions to exclude rather than include.

I didn't update `app.ini.sample` because I thought it migth be confusing for the user, since the default value is empty, anyway.

This PR is _related_ to #4831.

### CODE REVIEW

**Comment by lafriks** [comment]
Might it not be better to set add extension not to index then?

**Comment by guillep2k** [comment]
I suppose it depends on your particular case. I manage my company repository and there are projects ranging from VB6, C, C++, VB.Net, PHP, to Java and embedded. I still can't decide which method is more relevant to us. Lss, map, and sym are examples of files that definitely get in my way, but I'll probably decide for a list of known source code extensions to include rather than the opposite. There might be a lot of other use cases I can't think of, however.
I think that the **REPO_EXTENSIONS_LIST_EXCLUDE** flag is definitely useful, but there could be a **REPO_EXTENSIONS_LIST_INCLUDE** flag in its stead. The reason I decided for the former is that I could explain it better in fewer words 😅.

**Comment by lafriks** [comment]
Include list would be usually more hard to maintain and you will most probably don't even know if you are missing something that is not indexed. Include list would probably have very rare use case imho

**Comment by guillep2k** [comment]
Would you rather have a **REPO_EXTENSIONS_LIST_INCLUDE** flag, then?

**Comment by lafriks** [comment]
Yes that would be less confusing

**Comment by guillep2k** [comment]
OK, syntax is now as follows:
```
[Indexer]
...
REPO_INDEXER_EXTENSIONS = go, *.c, .cpp, h, idl, php, .
REPO_EXTENSIONS_LIST_INCLUDE = false
```
**REPO_INDEXER_EXTENSIONS** (default: empty) is a comma separated list of file extensions to **exclude** from the index. An empty list means _do not exclude files_ (i.e. the current behaviour).
**REPO_EXTENSIONS_LIST_INCLUDE** (default: `false`) if true, **REPO_INDEXER_EXTENSIONS** becomes a list of extensions to **include** rather than **exclude**.

**REPO_INDEXER_EXTENSIONS** now accepts `*.txt`, `.txt` and `txt` as synonyms. Dot (`.`) matches files with no extension.

**Comment by lafriks** [comment]
Probably you could use https://github.com/gobwas/glob for wildcard support

**Comment by guillep2k** [comment]
I've just added the `gobwas/glob` library for wildcard support. However, I think that power comes at the cost of added complexity.

In order to profit from the power of the glob patterns I'm using the full file path for matching (i.e. `lib/drivers/uart.c` vs. `uart.c`). This means patterns are not as intuitive as they could be. For instance, to match all the text files you need to use `**.txt` because the pattern `*.txt` only matches the files in the top level dir. On the other hand, it's possible to filter entire sub-directories using expressions like `data/**, **/data/**`.

I could have ignored the directory part of the path in the match and concentrate only on the actual file name, but then the library would be underutilized. Adding some examples to the docs would be very useful. I didn't want to pollute the `config-cheat-sheet` too much, so I kept it simple there.

To better reflect the effect of the filter, I've changed the names of the settings to `REPO_INDEXER_PATTERNS` and `REPO_PATTERNS_INCLUDE`.

**Note:** the last commit won't pass the build test because I don't know how to proceed with the added vendor. I've pushed just the code for your consideration.

PS: Sorry about the commit mess.

File: models/repo_indexer.go [hunk 1]
```diff
@@ -232,6 +232,19 @@ func addDelete(filename string, repo *Repository, batch rupture.FlushingBatch) e
...
+			if g.Match(name) {
+				found = true
+				break
+			}
+		}
+		if found != setting.Indexer.IncludePatterns {
+			return false
+		}
+	}
 	return entry.IsRegular() || entry.IsExecutable()
```
Comment (lafriks) [inline models/repo_indexer.go]: this check should probably go before file pattern check. Also `if setting.Indexer.FilePatterns != nil {` should be reversed for cleaner code, something like:

```go
func isIndexable(entry *git.TreeEntry) bool {
	if !entry.IsRegular() && !entry.IsExecutable() {
		return false
	}
	if setting.Indexer.FilePatterns == nil {
		return true
	}
	// Check pattern match
	var found bool
	...
```

Comment (guillep2k) [reply models/repo_indexer.go]: OK, I'll turn that the other way around. About the new vendor, ¿should I simply "make vendor" and add anything that's modified in a new commit?

Comment (guillep2k) [reply models/repo_indexer.go]: Done!

**Comment by lafriks** [comment]
why not use two patterns (include and exclude), so that it would be more flexible?

**Comment by guillep2k** [comment]
> 
> 
> why not use two patterns (include and exclude), so that it would be more flexible?

@lafriks No problem. Back when it was only filtering by simple patterns it was not useful, but now with a glob library it makes sense. I'll work on it.

**Comment by guillep2k** [comment]
@lafriks What do you think should take precedence, include or exclude?

**Comment by guillep2k** [comment]
I propose exclude should have precedence over include. Example scenario:
```
REPO_INDEXER_INCLUDE = /project, /resources
REPO_INDEXER_EXCLUDE = .gitignore
```

**Comment by lafriks** [comment]
Yes I agree that exclude should be taken into account first

**Comment by guillep2k** [comment]
Added unit and integration tests. Please notice that the `gobwas/glob` library has a particular syntax, so my previous example:
```
REPO_INDEXER_INCLUDE = /project, /resources
REPO_INDEXER_EXCLUDE = .gitignore
```
should have been in fact written like:
```
REPO_INDEXER_INCLUDE = /project/**, /resources/**
REPO_INDEXER_EXCLUDE = **.gitignore
```

**Review by lunny — APPROVED** [review]

File: go.mod [hunk 2]
```diff
@@ -47,6 +47,7 @@ require (
 	github.com/go-sql-driver/mysql v1.4.1
 	github.com/go-swagger/go-swagger v0.20.1
 	github.com/go-xorm/xorm v0.7.7-0.20190822154023-17592d96b35b
+	github.com/gobwas/glob v0.2.3
```
Comment (typeless) [inline go.mod]: ~~Why does it have to be updated?~~
Never mind. I got it.

Comment (typeless) [reply go.mod]: is it because it is used in the tests?

**Review by typeless — APPROVED** [review]

**Comment by guillep2k** [comment]
As this PR mutated quite a bit along the way, here is a summary of its functionality.

## What this PR does

Adds an option in `app.ini` to restrict _repository indexing_ by glob pattern matching using the `gobwas/glob` library:
```
[Indexer]
...
REPO_INDEXER_INCLUDE = project/**, resources/**
REPO_INDEXER_EXCLUDE = **.gitignore
```

It can help when indexes become polluted with derived files (e.g. `lss`, `sym`, `map`, etc.) so the search results are more relevant. It can also help reduce the index size.

**REPO_INDEXER_INCLUDE** (default: empty) is a comma separated list of glob patterns to **include** in the index. An empty list means "include all files".
**REPO_INDEXER_EXCLUDE** (default: empty) is a comma separated list of glob patterns to **exclude** from the index. Files that match this list will not be indexed. `REPO_INDEXER_EXCLUDE` takes precedence over `REPO_INDEXER_INCLUDE`.

Glob patterns as implemented by `gobwas/glob` are powerful and very fast, albeit a little complex. They work as follows:

* To match all files with .txt extension, no matter what directory, use `**.txt`.
* To match all files with .txt extension at the root level of the repository, use `*.txt`.
* To match all files inside `resources/bin` and below, use `resources/bin/**`.
* To match all files immediately inside `resources/bin`, use `resources/bin/*`.
* To match all Makefile files, use `**Makefile`.
* Matching a directory has no effect; `resources/bin` will not include/exclude files inside that directory; `resources/bin/**` will.
* All files and patterns are normalized to lower case.

File: modules/setting/indexer.go [hunk 3]
```diff
@@ -51,10 +58,32 @@ func newIndexerService() {
...
+		expr = strings.TrimSpace(expr)
+		if expr != "" {
+			if g, err := glob.Compile(expr, '.', '/'); err != nil {
+				log.Info("Invalid glob expresion '%s' (skipped): %v", expr, err)
+			} else {
+				extarr = append(extarr, g)
+			}
+		}
+	}
+	if len(extarr) == 0 {
```
Comment (lafriks) [inline modules/setting/indexer.go]: imho this is not needed

File: models/repo_indexer.go [hunk 4]
```diff
@@ -234,7 +234,24 @@ func addDelete(filename string, repo *Repository, batch rupture.FlushingBatch) e
...
+	if !entry.IsRegular() && !entry.IsExecutable() {
+		return false
+	}
+	name := strings.ToLower(entry.Name())
+	for _, g := range setting.Indexer.ExcludePatterns {
+		if g.Match(name) {
+			return false
+		}
+	}
+	if setting.Indexer.IncludePatterns == nil {
```
Comment (lafriks) [inline models/repo_indexer.go]: This is also not needed

File: models/repo_indexer.go [hunk 5]
```diff
@@ -234,7 +234,24 @@ func addDelete(filename string, repo *Repository, batch rupture.FlushingBatch) e
...
+		return true
+	}
+	for _, g := range setting.Indexer.IncludePatterns {
+		if g.Match(name) {
+			return true
+		}
+	}
+	return false
```
Comment (lafriks) [inline models/repo_indexer.go]:
```suggestion
	return len(setting.Indexer.IncludePatterns) == 0
```

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by lafriks** [comment]
these last small things but otherwise great job! 🎉

File: models/repo_indexer.go (hunk 5 above)
Comment (guillep2k) [reply models/repo_indexer.go]: Done * 3

**Review by lafriks — APPROVED** [review]

**Comment by guillep2k** [comment]
Now gobwas is one of the vendors, so it makes more sense to use it here.

**Comment by lafriks** [comment]
make lg-tm work

**Pull request merged**